solidity/contracts/TokenStaking.sol

```diff
@@ -107,44 +107,166 @@ contract TokenStaking is Authorizations, StakeDelegatable {
         return MinimumStakeSchedule.current(minimumStakeScheduleStart);
     }
 
-    /// @notice Receives approval of token transfer and stakes the approved amount.
-    /// @dev Makes sure provided token contract is the same one linked to this contract.
+    /// @notice Receives approval of token transfer and stakes the approved
+    /// amount or top-ups an existing delegation with the approved amount.
+    /// In case of a top-up, it is expected that the operator stake passed
+    /// initialization period, it is not undelegated and that the top-up is
+    /// performed from the same source of tokens as the initial delegation.
+    /// That is, if the tokens were delegated from a grant, top-up
+    /// has to be performed from the same grant. If the delegation was done
+    /// using liquid tokens, only liquid tokens from the same owner can be used
+    /// to top-up the stake.
+    /// @dev Makes sure provided token contract is the same one linked to this
+    /// contract.
     /// @param _from The owner of the tokens who approved them to transfer.
     /// @param _value Approved amount for the transfer and stake.
     /// @param _token Token contract address.
     /// @param _extraData Data for stake delegation. This byte array must have
     /// the following values concatenated:
-    /// - Beneficiary address (20 bytes)
+    /// - Beneficiary address (20 bytes), ignored for a top-up
     /// - Operator address (20 bytes)
-    /// - Authorizer address (20 bytes)
-    function receiveApproval(address _from, uint256 _value, address _token, bytes memory _extraData) public {
+    /// - Authorizer address (20 bytes), ignored for a top-up
+    /// - Grant ID (32 bytes) - required only when called by TokenStakingEscrow
+    function receiveApproval(
+        address _from,
+        uint256 _value,
+        address _token,
+        bytes memory _extraData
+    ) public {
         require(ERC20Burnable(_token) == token, "Unrecognized token contract");
         require(_value >= minimumStake(), "Value must be greater than the minimum stake");
```

Comment:
The minimum value for a top up has to be at least equal to the minimum stake, this seems not necessary as the minimum amount is already staked.

Reply:
Agreed. Please let me deal with it once https://github.com/keep-network/keep-core/pull/1902 and https://github.com/keep-network/keep-core/issues/1883 are there. We'll have to fight for every single byte because of out-of-gas problems during deployment.

Reply:
Covered in https://github.com/keep-network/keep-core/pull/1974